File: src/services/customers.ts
import { ApiError, jsonFetch } from "./http";
import { getStoredToken } from "./auth";

type UUID = string;

type RequestOptions = {
  token?: string | null;
  signal?: AbortSignal;
};

type CursorResponse = {
  total: number;
  hasNext: boolean;
  nextCursor: string | null;
};

export type CustomerListItem = {
  id: UUID;
  fullName: string;
  primaryEmail: string | null;
  conversationsCount: number;
  satisfactionScore: number | null;
  lastContactAt: string | null;
  lifecycleStage: string;
};

export type CustomersListResponse = CursorResponse & {
  items: CustomerListItem[];
};

export type CustomerStats = {
  conversationsTotal: number;
  conversationsLast30Days: number;
  csatAverage: number | null;
  csatTrend: number | null;
  expansionOpportunities: number;
};

export type CustomerContactMethod = {
  type: "email" | "phone" | "social" | "messenger" | "other";
  value: string;
  isPrimary: boolean;
};

export type CustomerTag = {
  id: UUID;
  label: string;
  color: string | null;
};

export type CustomerNote = {
  id: UUID;
  customerId: UUID;
  authorUserId: UUID | null;
  authorAgentId: UUID | null;
  visibility: "internal" | "shared";
  body: string;
  pinned: boolean;
  createdAt: string;
  updatedAt: string;
};

export type CustomerActivityEvent = {
  id: UUID;
  customerId: UUID;
  eventType: string;
  occurredAt: string;
  actorUserId: UUID | null;
  actorAgentId: UUID | null;
  actorCustomerId: UUID | null;
  caseId: UUID | null;
  conversationId: UUID | null;
  details: Record<string, unknown> | null;
};

export type CustomerCaseLink = {
  id: UUID;
  title: string;
  status: string;
  priority: string;
  openedAt: string;
};

export type CustomerDetail = {
  id: UUID;
  businessId: UUID;
  fullName: string;
  primaryEmail: string | null;
  primaryPhone: string | null;
  country: string | null;
  lifecycleStage: string;
  satisfactionScore: number | null;
  personaTags: string[];
  lastContactAt: string | null;
  createdAt: string;
  updatedAt: string;
  stats: CustomerStats;
  contacts: CustomerContactMethod[];
  tags: CustomerTag[];
  casesOpen: CustomerCaseLink[];
  casesResolved: CustomerCaseLink[];
  notes: CustomerNote[];
  activity: CustomerActivityEvent[];
};

export type CustomerDetailResponse = {
  customer: CustomerDetail;
};

export type CustomerDetailInclude = Array<"notes" | "activity" | "cases">;

export type CustomersListParams = {
  search?: string;
  lifecycleStage?: string;
  tags?: string[];
  dateFrom?: string;
  dateTo?: string;
  limit?: number;
  cursor?: string | null;
};

export type CustomerCreateRequest = {
  fullName: string;
  primaryEmail?: string | null;
  primaryPhone?: string | null;
  country?: string | null;
  lifecycleStage: string;
  satisfactionScore?: number | null;
  personaTags?: string[];
  contactMethods?: CustomerContactMethod[];
  tagIds?: UUID[];
};

export type CustomerUpdateRequest = Partial<CustomerCreateRequest>;

export type CustomerNoteCreateRequest = {
  body: string;
  visibility: "internal" | "shared";
  pinned?: boolean;
};

export type CustomerNoteUpdateRequest = {
  body?: string;
  visibility?: "internal" | "shared";
  pinned?: boolean;
};

export type CustomerImportRow = {
  fullName: string;
  email?: string | null;
  phone?: string | null;
  country?: string | null;
  lifecycleStage?: string | null;
  tags?: string[];
};

export type CustomerImportRequest = {
  rows: CustomerImportRow[];
  skipDuplicates?: boolean;
};

export type CustomerImportResult = {
  importedCount: number;
  skippedCount: number;
  errors: string[];
};

export type CustomerNotesResponse = CursorResponse & {
  items: CustomerNote[];
};

export type CustomerActivityResponse = CursorResponse & {
  items: CustomerActivityEvent[];
};

const buildQuery = (params: CustomersListParams | undefined) => {
  if (!params) return "";
  const searchParams = new URLSearchParams();
  if (params.search) searchParams.set("search", params.search);
  if (params.lifecycleStage) searchParams.set("lifecycle_stage", params.lifecycleStage);
  if (params.dateFrom) searchParams.set("date_from", params.dateFrom);
  if (params.dateTo) searchParams.set("date_to", params.dateTo);
  if (typeof params.limit === "number") searchParams.set("limit", String(params.limit));
  if (params.cursor) searchParams.set("cursor", params.cursor);
  if (Array.isArray(params.tags)) {
    params.tags.filter(Boolean).forEach((tag) => searchParams.append("tags", tag));
  }
  const query = searchParams.toString();
  return query ? `?${query}` : "";
};

const withAuth = (token?: string | null) => token ?? getStoredToken();

export async function listCustomers(
  params?: CustomersListParams,
  options: RequestOptions = {}
): Promise<CustomersListResponse> {
  const query = buildQuery(params);
  try {
    return await jsonFetch<CustomersListResponse>(`/v1/customers${query}`, {
      method: "GET",
      token: withAuth(options.token),
      signal: options.signal,
    });
  } catch (error) {
    if (error instanceof ApiError && (error.code === "configuration_error" || error.code === "network_error")) {
      if (import.meta.env?.DEV) {
        console.warn("[customers] Falling back to empty list due to API error", error.message);
      }
      return {
        items: [],
        total: 0,
        hasNext: false,
        nextCursor: null,
      };
    }
    throw error;
  }
}

export async function getCustomerDetail(
  customerId: UUID,
  include?: CustomerDetailInclude,
  options: RequestOptions = {}
): Promise<CustomerDetailResponse> {
  const searchParams = new URLSearchParams();
  include?.forEach((value) => searchParams.append("include", value));
  const suffix = searchParams.size ? `?${searchParams.toString()}` : "";
  return jsonFetch<CustomerDetailResponse>(`/v1/customers/${customerId}${suffix}`, {
    method: "GET",
    token: withAuth(options.token),
    signal: options.signal,
  });
}

export async function createCustomer(
  payload: CustomerCreateRequest,
  options: RequestOptions = {}
): Promise<CustomerDetailResponse> {
  return jsonFetch<CustomerDetailResponse>("/v1/customers", {
    method: "POST",
    body: payload,
    token: withAuth(options.token),
    signal: options.signal,
  });
}

export async function updateCustomer(
  customerId: UUID,
  payload: CustomerUpdateRequest,
  options: RequestOptions = {}
): Promise<CustomerDetailResponse> {
  return jsonFetch<CustomerDetailResponse>(`/v1/customers/${customerId}`, {
    method: "PATCH",
    body: payload,
    token: withAuth(options.token),
    signal: options.signal,
  });
}

export async function listCustomerNotes(
  customerId: UUID,
  params?: { limit?: number; cursor?: string | null },
  options: RequestOptions = {}
): Promise<CustomerNotesResponse> {
  const searchParams = new URLSearchParams();
  if (typeof params?.limit === "number") searchParams.set("limit", String(params.limit));
  if (params?.cursor) searchParams.set("cursor", params.cursor);
  const suffix = searchParams.size ? `?${searchParams.toString()}` : "";
  return jsonFetch<CustomerNotesResponse>(`/v1/customers/${customerId}/notes${suffix}`, {
    method: "GET",
    token: withAuth(options.token),
    signal: options.signal,
  });
}

export async function createCustomerNote(
  customerId: UUID,
  payload: CustomerNoteCreateRequest,
  options: RequestOptions = {}
): Promise<CustomerNote> {
  return jsonFetch<CustomerNote>(`/v1/customers/${customerId}/notes`, {
    method: "POST",
    body: payload,
    token: withAuth(options.token),
    signal: options.signal,
  });
}

export async function updateCustomerNote(
  customerId: UUID,
  noteId: UUID,
  payload: CustomerNoteUpdateRequest,
  options: RequestOptions = {}
): Promise<CustomerNote> {
  return jsonFetch<CustomerNote>(`/v1/customers/${customerId}/notes/${noteId}`, {
    method: "PATCH",
    body: payload,
    token: withAuth(options.token),
    signal: options.signal,
  });
}

export async function listCustomerActivity(
  customerId: UUID,
  params?: { limit?: number; cursor?: string | null },
  options: RequestOptions = {}
): Promise<CustomerActivityResponse> {
  const searchParams = new URLSearchParams();
  if (typeof params?.limit === "number") searchParams.set("limit", String(params.limit));
  if (params?.cursor) searchParams.set("cursor", params.cursor);
  const suffix = searchParams.size ? `?${searchParams.toString()}` : "";
  return jsonFetch<CustomerActivityResponse>(`/v1/customers/${customerId}/activity${suffix}`, {
    method: "GET",
    token: withAuth(options.token),
    signal: options.signal,
  });
}

export async function importCustomers(
  payload: CustomerImportRequest,
  options: RequestOptions = {}
): Promise<CustomerImportResult> {
  return jsonFetch<CustomerImportResult>("/v1/customers/import", {
    method: "POST",
    body: payload,
    token: withAuth(options.token),
    signal: options.signal,
  });
}

File: backend/app/api/v1/schemas/customers.py
"""API-layer schemas for Customer 360 endpoints."""

from __future__ import annotations

from datetime import datetime, date
from typing import Literal, Sequence
from uuid import UUID

from pydantic import BaseModel, ConfigDict, EmailStr, Field


def _to_camel(value: str) -> str:
    head, *tail = value.split("_")
    return head + "".join(token.capitalize() for token in tail)


class CamelModel(BaseModel):
    """Base model that maps snake_case fields to camelCase aliases."""

    model_config = ConfigDict(populate_by_name=True, alias_generator=_to_camel)


class CustomerListItem(CamelModel):
    id: UUID
    full_name: str
    primary_email: EmailStr | None = None
    conversations_count: int
    satisfaction_score: float | None = Field(default=None, ge=0, le=100)
    last_contact_at: datetime | None = None
    lifecycle_stage: str


class CustomerStats(CamelModel):
    conversations_total: int
    conversations_last_30_days: int
    csat_average: float | None = Field(default=None, ge=0, le=100)
    csat_trend: float | None = Field(default=None, ge=-100, le=100)
    expansion_opportunities: int


class CustomerContactMethod(CamelModel):
    type: Literal["email", "phone", "social", "messenger", "other"]
    value: str = Field(min_length=2, max_length=255)
    is_primary: bool = False


class CustomerTag(CamelModel):
    id: UUID
    label: str
    color: str | None = Field(default=None, pattern=r"^[A-Fa-f0-9]{6}$")


class CustomerNote(CamelModel):
    id: UUID
    customer_id: UUID
    author_user_id: UUID | None = None
    author_agent_id: UUID | None = None
    visibility: Literal["internal", "shared"]
    body: str
    pinned: bool
    created_at: datetime
    updated_at: datetime


class CustomerActivityEvent(CamelModel):
    id: UUID
    customer_id: UUID
    event_type: str
    occurred_at: datetime
    actor_user_id: UUID | None = None
    actor_agent_id: UUID | None = None
    actor_customer_id: UUID | None = None
    case_id: UUID | None = None
    conversation_id: UUID | None = None
    details: dict | None = None


class CustomerCaseLink(CamelModel):
    id: UUID
    title: str
    status: str
    priority: str
    opened_at: datetime


class CustomerDetail(CamelModel):
    id: UUID
    business_id: UUID
    full_name: str
    primary_email: EmailStr | None = None
    primary_phone: str | None = None
    country: str | None = None
    lifecycle_stage: str
    satisfaction_score: float | None = Field(default=None, ge=0, le=100)
    persona_tags: list[str]
    last_contact_at: datetime | None = None
    created_at: datetime
    updated_at: datetime
    stats: CustomerStats
    contacts: list[CustomerContactMethod]
    tags: list[CustomerTag]
    cases_open: list[CustomerCaseLink]
    cases_resolved: list[CustomerCaseLink]
    notes: list[CustomerNote]
    activity: list[CustomerActivityEvent]


class CustomersListResponse(CamelModel):
    items: list[CustomerListItem]
    total: int
    has_next: bool
    next_cursor: str | None = None


class CustomerDetailResponse(CamelModel):
    customer: CustomerDetail


class CustomerActivityListResponse(CamelModel):
    items: list[CustomerActivityEvent]
    total: int
    has_next: bool
    next_cursor: str | None = None


class CustomerNoteListResponse(CamelModel):
    items: list[CustomerNote]
    total: int
    has_next: bool
    next_cursor: str | None = None


class CustomersListQuery(CamelModel):
    search: str | None = Field(default=None, max_length=120)
    lifecycle_stage: str | None = None
    tags: list[str] | None = None
    date_from: date | None = None
    date_to: date | None = None
    limit: int = Field(default=25, ge=1, le=100)
    cursor: str | None = Field(default=None, max_length=120)


class CustomerCreateRequest(CamelModel):
    full_name: str = Field(min_length=1, max_length=160)
    primary_email: EmailStr | None = None
    primary_phone: str | None = Field(default=None, max_length=64)
    country: str | None = Field(default=None, min_length=2, max_length=2)
    lifecycle_stage: str
    satisfaction_score: float | None = Field(default=None, ge=0, le=100)
    persona_tags: list[str] = Field(default_factory=list, max_length=20)
    contact_methods: list[CustomerContactMethod] = Field(default_factory=list, max_length=10)
    tag_ids: list[UUID] = Field(default_factory=list, max_length=30)


class CustomerUpdateRequest(CamelModel):
    full_name: str | None = Field(default=None, min_length=1, max_length=160)
    primary_email: EmailStr | None = None
    primary_phone: str | None = Field(default=None, max_length=64)
    country: str | None = Field(default=None, min_length=2, max_length=2)
    lifecycle_stage: str | None = None
    satisfaction_score: float | None = Field(default=None, ge=0, le=100)
    persona_tags: list[str] | None = Field(default=None, max_length=20)
    contact_methods: list[CustomerContactMethod] | None = Field(default=None, max_length=10)
    tag_ids: list[UUID] | None = Field(default=None, max_length=30)


class CustomerNoteCreateRequest(CamelModel):
    body: str = Field(min_length=1, max_length=4000)
    visibility: Literal["internal", "shared"]
    pinned: bool = False


class CustomerNoteUpdateRequest(CamelModel):
    body: str | None = Field(default=None, min_length=1, max_length=4000)
    visibility: Literal["internal", "shared"] | None = None
    pinned: bool | None = None


class CustomerImportRow(CamelModel):
    full_name: str = Field(min_length=1, max_length=160)
    email: EmailStr | None = None
    phone: str | None = Field(default=None, max_length=64)
    country: str | None = Field(default=None, min_length=2, max_length=2)
    lifecycle_stage: str | None = None
    tags: list[str] = Field(default_factory=list)


class CustomerImportRequest(CamelModel):
    rows: list[CustomerImportRow] = Field(min_length=1, max_length=1000)
    skip_duplicates: bool = True


class CustomerImportResult(CamelModel):
    imported_count: int
    skipped_count: int
    errors: list[str]


class CustomerTagCreateRequest(CamelModel):
    label: str = Field(min_length=1, max_length=64)
    color: str | None = Field(default=None, pattern=r"^[A-Fa-f0-9]{6}$")


class CustomerTagAssignmentRequest(CamelModel):
    tag_ids: Sequence[UUID]


class CustomerTagsResponse(CamelModel):
    items: list[CustomerTag]


__all__ = [
    "CamelModel",
    "CustomerActivityEvent",
    "CustomerActivityListResponse",
    "CustomerCaseLink",
    "CustomerContactMethod",
    "CustomerCreateRequest",
    "CustomerDetail",
    "CustomerDetailResponse",
    "CustomerImportRequest",
    "CustomerImportResult",
    "CustomerImportRow",
    "CustomerListItem",
    "CustomerNote",
    "CustomerNoteCreateRequest",
    "CustomerNoteListResponse",
    "CustomerNoteUpdateRequest",
    "CustomerStats",
    "CustomerTag",
    "CustomerTagAssignmentRequest",
    "CustomerTagCreateRequest",
    "CustomerTagsResponse",
    "CustomerUpdateRequest",
    "CustomersListQuery",
    "CustomersListResponse",
]
